### DISCUSSION TITLE
Equivalent for mapboxGL `flyTo` camera transition?

### DISCUSSION THREAD

**Discussion by benschlegel** [visgl/react-google-maps#104] [General]
In visgl's other repository ([react-map-gl](https://github.com/visgl/react-map-gl/)) theres an example for [camera transitions](https://github.com/visgl/react-map-gl/tree/7.1-release/examples/viewport-animation). This internally uses the `.flyTo` function of mapbox, which there is no direct equivalent with google maps as far as i can tell.

Does anyone know of a way to achieve a similar effect to make the camera zoom out, pan and zoom back in as smoothly as in the `.flyTo` example?

**Comment by ibgreen** [comment]
FWIW - deck.gl has a FlyToInterpolator that handles this for deck.gl applications (deck can also control and animate base maps). It could be worth testing that this works with the google basemap. I suspect the FlyToInterpolator code could also be deconstructed and rebuilt as a (deck.gl-independent) part of this  react-google-maps library for those who want to use google-maps standalone without deck.gl

**Reply by benschlegel** [reply]
Thanks for the help! Are there any examples of how deck.gl integrates with react-google-maps directly (if possible at all)? 

I've found this example on the deck.gl docs for `ViewTransitions`: https://deck.gl/docs/developer-guide/view-state-transitions#usage
Can I just expect to switch `<StaticMap>`  from MapBox/react-map-gl for the `<Map>` component from this library and it should work?

**Reply by usefulthink** [reply]
The documentation on integrating with deck.gl is still a bit sparse and well hidden, but you can find it here: https://visgl.github.io/react-google-maps/docs/api-reference/components/map#deckgl-integration

**Reply by usefulthink** [reply]
And I looked it up, here are some pointers to the code used in deck.gl:
 - [`FlyToInterpolator`](https://github.com/visgl/deck.gl/blob/master/modules/core/src/transitions/fly-to-interpolator.ts)
 - [`flyToViewport()` function in math.gl](https://github.com/uber-web/math.gl/blob/master/modules/web-mercator/src/fly-to-viewport.ts)
 - and this appears to be the paper that describes the math behind it: https://www.win.tue.nl/~vanwijk/zoompan.pdf

**Reply by benschlegel** [reply]
Thanks a lot, that's really helpful!

**Reply by benschlegel** [reply]
Hey again, not sure if this is the right place to ask, please let me know if not. I was trying to use deck.gl as a minimal wrapper to be able to use its `ViewTransition` feature with this library like this:

```jsx
<DeckGL
	  initialViewState={initialViewState}
	  controller={true}
	  onViewStateChange={limitTiltRange}
	  layers={layers}
>
	<Map {...props />
</DeckGL>
```

While this does appear to work at first (changing lat/lng on initialViewState also changes it for the underlying map), the problem that i'm now running into is that the deck.gl overlay (where theres nothing being rendered anywhere) blocks all clicks/hovers from being passed through to the map. 

### The Problem:

I'm no longer able to click my google maps custom/built in controls and marker click events dont fire anymore. I can still pan/zoom the map and double click to zoom also still works (I also think for my use-case i dont need the layers as I'm only trying to use it for the view transition, but this behaviour occurs with and without specifying the layers property).

I dont think this is the main problem, but for completeness sake, I get the following error in my console:
```
 ⚠ ./node_modules/@loaders.gl/worker-utils/dist/es5/lib/node/require-utils.node.js
Critical dependency: the request of a dependency is an expression

Import trace for requested module:
./node_modules/@loaders.gl/worker-utils/dist/es5/lib/node/require-utils.node.js
./node_modules/@loaders.gl/worker-utils/dist/es5/lib/library-utils/library-utils.js
./node_modules/@loaders.gl/worker-utils/dist/es5/index.js
./node_modules/@loaders.gl/core/dist/es5/lib/api/parse.js
./node_modules/@loaders.gl/core/dist/es5/index.js
./node_modules/@deck.gl/layers/dist/es5/icon-layer/icon-manager.js
./node_modules/@deck.gl/layers/dist/es5/icon-layer/icon-layer.js
./node_modules/@deck.gl/layers/dist/es5/index.js
./src/components/Map.tsx
./src/app/page.tsx
```

I'm using next.js for my project and am importing my `Map` component like described in the deck.gl docs here: https://deck.gl/docs/get-started/using-with-react#using-deckgl-with-ssr (which I think is supposed to solve this problem so I don't think it's the root cause).

**Reply by usefulthink** [reply]
This is probably the reason: https://github.com/visgl/react-google-maps/blob/main/src/components/map/index.tsx#L88-L89

It still feels like a hack to me and I'm working on improving this part specifically. When deck.gl is supposed to be in control, we need to prevent google maps from trying to update it's own camera which is difficult and right now achieved by just moving it to the back...

**Reply by benschlegel** [reply]
That does sound like a really difficult problem. I guess there's no real workaround until the core issue is fixed?

I tried manually increasing the maps z-index, which does somewhat work and also makes the view transition work, but as you mentioned, the camera quickly starts breaking as soon as i change zoom level/fit to bounds and pan, the camera starts jumping around between what I assume is google maps managed camera state and deck.gl managed camera state.

**Reply by usefulthink** [reply]
you could try using the `onBoundsChanged` event to get notified of every change the map instance does itself and put that into the  viewstate of deck.gl, but I suspect this might cause some stuttering.

**Reply by benschlegel** [reply]
this somewhat seems to behave as expected, but the entire animation/transition is extremely choppy compared to the [mapbox-gl example](https://visgl.github.io/react-map-gl/examples/viewport-animation). 

is this something to be expected/a limitation of google maps tile fetching/rendering? or more likely to be an issue of how i use deck.gl/this library

**Reply by usefulthink** [reply]
It's not you, it's us :D

Synchronizing deck.gl and the basemap is notoriously difficult if you want both of them to be in control. If behaves decently if either google maps is prevented from updating it's state in any way, so the sole control lies with deck.gl or if you use deck.gl as an overlay and leave full control to the google maps instance.

It's also something I am still working on to figure out the best way to handle it, so maybe there are ways to still improve it.

`react-map-gl` has had years to figure this one out (and mapbox itself seems to be a bit more malleable in this regard) :)

**Comment by usefulthink** [comment]
I think this could be an interesting Idea for this library and it would be worthwhile to explore how this could fit into the reactive programming model.

The Google Maps API itself doesn't provide anything like this. There sometimes are automatic minimal camera-transitions when using `map.setCenter()` and so on (which is also why we're not using them for the center/zoom/heading/tilt props), but I think they are only triggered when moving smaller distances.

**Comment by dragonlobster** [comment]
how about `map.panTo()`?

**Comment by irg1008** [comment]
And how does google itself have that zoom out then zoom in animation on bounds change? Works with android sdj too

**Comment by kabir-afk** [comment]
Hey , I encountered the same problem and I was able to kinda create a transition of my own .Built the core zoom in logic using a combination of **recursion** ,`setTimeout`  and other map properties like `panto()`,`getZoom()` , `setZoom()`  , then refactored with AI to account for zoom out and in simultaneously. Below is my code :
```
import { useMap } from "@vis.gl/react-google-maps";

export function PanToButton({ position}) {
  const map = useMap();
  function zoomOutAndIn(event) {
    event.preventDefault();
    let minZoom = 3;
    let maxZoom = 8;
    // Step 1: Zoom out to minZoom
    function zoomOutStep() {
      if (map.getZoom() > minZoom) {
        map.setZoom(map.getZoom() - 1);
        setTimeout(zoomOutStep, 250);
      } else {
        // Step 2: Pan to position and zoom in to maxZoom
        map.panTo(position);
        setTimeout(zoomInStep, 250);
      }
    }
    function zoomInStep() {
      if (map.getZoom() < maxZoom) {
        map.setZoom(map.getZoom() + 1);
        setTimeout(zoomInStep, 250);
      }
    }
    zoomOutStep();
  }
  function panToPosition(event) {
    event.preventDefault();
    zoomOutAndIn(event);
  }
  return (
    <button
      type="submit"
      disabled={cursor}
      onClick={panToPosition}
    >
      GUESS
    </button>
  );
}

```

When someone clicks on the button , it'll zoom out to 3(or whatever you set `minZoom` to) , pan to the `position` passed in prop and then zoom in .The `PanToButton` has to be a child of `<Map/>`  to work , hope this helps someone out in the future. No doubt the animation is a bti chalky but good for a work around I'd say

**Reply by kabir-afk** [reply]
oh btw , I have no clue how deck.gl works alongside this library . This implementation was purely in react-google-maps
